cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh

```diff
@@ -0,0 +1,488 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_scan.cuh>
+#include <cub/block/block_store.cuh>
+#include <cub/block/radix_rank_sort_operations.cuh>
+
+#include <raft/cudart_utils.h>
+#include <raft/device_atomics.cuh>
+
+/*
+  Two implementations:
+
+  (1) radix select (select + filter):
+      first select the k-th value by going through radix passes,
+      then filter out all wanted data from original data
+
+  (2) radix topk:
+      filter out wanted data directly while going through radix passes
+*/
+
+namespace raft::spatial::knn::detail::ivf_flat {
+
+constexpr int BLOCK_DIM       = 512;
+constexpr int ITEM_PER_THREAD = 32;
+
+template <int BITS_PER_PASS>
+__host__ __device__ constexpr int calc_num_buckets()
+{
+  return 1 << BITS_PER_PASS;
+}
+
+template <typename T, int BITS_PER_PASS>
+__host__ __device__ constexpr int calc_num_passes()
+{
+  return (sizeof(T) * 8 - 1) / BITS_PER_PASS + 1;
+}
+
+// bit 0 is the least significant (rightmost) bit
+// this function works even when pass=-1, which is used in calc_mask()
+template <typename T, int BITS_PER_PASS>
+__device__ constexpr int calc_start_bit(int pass)
```

Comment:
I've had a feeling that both versions are common. I guess, here we go from the most to least significant bits, because this way we can save a few passes since, as you said, we don't need to sort within the `k`-sized buckets.